node/pkg/watchers/ibc/watcher.go

```diff
@@ -0,0 +1,796 @@
+	ID uint64 `json:"id"`
+}
+
+const (
+	// The IBC receiver contract publishes wasm events, not execute events.
+	contractAddressFilterKey = "wasm._contract_address"
+	contractAddressLogKey    = "_contract_address"
+)
+
+// Run is the runnable for monitoring the IBC contract on wormchain.
+func (w *Watcher) Run(ctx context.Context) error {
+	w.logger = supervisor.Logger(ctx)
+
+	errC := make(chan error)
+	defer close(errC)
+
+	// Rebuild these from scratch every time the watcher restarts.
+	w.connectionMap = make(map[string]*connectionEntry)
+	w.chainMap = make(map[vaa.ChainID]*connectionEntry)
+
+	// Build our internal data structures based on the config passed in.
+	for idx, chainToMonitor := range w.connectionConfig {
+		if w.channelData[idx].ChainID != chainToMonitor.ChainID {
+			panic("channelData is not in sync with chainToMonitor") // This would be a program bug!
+		}
+
+		_, exists := w.connectionMap[chainToMonitor.ConnID]
+		if exists {
+			return fmt.Errorf("detected duplicate IBC connection: %v", chainToMonitor.ConnID)
+		}
+
+		_, exists = w.chainMap[chainToMonitor.ChainID]
+		if exists {
+			return fmt.Errorf("detected duplicate chainID: %v", chainToMonitor.ChainID)
+		}
+
+		ce := &connectionEntry{
+			connectionID: chainToMonitor.ConnID,
+			chainID:      chainToMonitor.ChainID,
+			chainName:    vaa.ChainID(chainToMonitor.ChainID).String(),
+			readiness:    common.MustConvertChainIdToReadinessSyncing(chainToMonitor.ChainID),
+			msgC:         w.channelData[idx].MsgC,
+			obsvReqC:     w.channelData[idx].ObsvReqC,
+		}
+
+		w.logger.Info("will monitor chain over IBC", zap.String("chain", ce.chainName), zap.String("IBC connection", ce.connectionID))
+		w.connectionMap[ce.connectionID] = ce
+		w.chainMap[ce.chainID] = ce
+
+		p2p.DefaultRegistry.SetNetworkStats(ce.chainID, &gossipv1.Heartbeat_Network{ContractAddress: w.contractAddress})
+	}
+
+	w.logger.Info("creating watcher", zap.String("wsUrl", w.wsUrl), zap.String("lcdUrl", w.lcdUrl), zap.String("contract", w.contractAddress))
+
+	c, _, err := websocket.Dial(ctx, w.wsUrl, nil)
+	if err != nil {
+		connectionErrors.WithLabelValues("websocket_dial_error").Inc()
+		return fmt.Errorf("failed to establish tendermint websocket connection: %w", err)
+	}
+	defer c.Close(websocket.StatusNormalClosure, "")
+
+	// This was copied from the cosmwasm watcher. It sure it's still necessary. . .
+	c.SetReadLimit(524288)
+
+	// Subscribe to smart contract transactions.
+	params := [...]string{fmt.Sprintf("tm.event='Tx' AND %s='%s'", contractAddressFilterKey, w.contractAddress)}
+	command := &clientRequest{
+		JSONRPC: "2.0",
+		Method:  "subscribe",
+		Params:  params,
+		ID:      1,
+	}
+	err = wsjson.Write(ctx, c, command)
+	if err != nil {
+		connectionErrors.WithLabelValues("websocket_subscription_error").Inc()
+		return fmt.Errorf("failed to subscribe to events: %w", err)
+	}
+
+	// Wait for the success response.
+	_, _, err = c.Read(ctx)
```

Comment:
should the value of the response be checked?

Reply:
Added a check.